packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,387 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { Script } from "forge-std/Script.sol";
+import { console2 as console } from "forge-std/console2.sol";
+
+import { Artifacts } from "scripts/Artifacts.s.sol";
+import { DeployConfig } from "scripts/DeployConfig.s.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { L1StandardBridge } from "src/L1/L1StandardBridge.sol";
+import { L1CrossDomainMessenger } from "src/L1/L1CrossDomainMessenger.sol";
+import { L2StandardBridge } from "src/L2/L2StandardBridge.sol";
+import { L2CrossDomainMessenger } from "src/L2/L2CrossDomainMessenger.sol";
+import { SequencerFeeVault } from "src/L2/SequencerFeeVault.sol";
+import { FeeVault } from "src/universal/FeeVault.sol";
+import { OptimismMintableERC20Factory } from "src/universal/OptimismMintableERC20Factory.sol";
+import { L1Block } from "src/L2/L1Block.sol";
+import { GovernanceToken } from "src/governance/GovernanceToken.sol";
+import { EIP1967Helper } from "test/mocks/EIP1967Helper.sol";
+
+interface IInitializable {
+    function initialize(address _addr) external;
+}
+
+/// @dev The general flow of adding a predeploy is:
+///      1. _setProxies uses vm.etch to set the Proxy.sol deployed bytecode for proxy address 0x420...0 to 0x420...799.
+///      Additionally, the PROXY_ADMIN_ADDRESS and PROXY_IMPLEMENTATION_ADDRESS storage slots are set for the proxy
+/// address.
+///      2. vm.etch sets the deployed bytecode for each predeploy at the implementation address (i.e. code namespace).
+///      3. The `initialize` method is called at the implementation address with zero/dummy vaules if it exists.
+///      4. The `initialize` method is called at the proxy address with actual vaules if it exists.
+///      5. A `require` check to verify the expected implementation address is set for the proxy.
+/// @notice The following safety invariants are used when setting state:
+///         1. `vm.getDeployedBytecode` can only be used with `vm.etch` when there are no side
+///         effects in the constructor and no immutables in the bytecode.
+///         2. A contract must be deployed using the `new` syntax if there are immutables in the code.
+///         Any other side effects from the init code besides setting the immutables must be cleaned up afterwards.
+///         3. A contract is deployed using the `new` syntax because it's not proxied, but still needs to be set
+///         at a specific address. Because just deploying a new instance doesn't give us the contract at our desired
+/// address,
```

Comment:
nit
```suggestion
///         at a specific address. Because just deploying a new instance doesn't give us the contract at our desired
///         address,
```

Reply:
Committed